odl/space/fspace.py

```diff
@@ -154,196 +356,346 @@ class FunctionSetVector(Operator):
 
     """Representation of a `FunctionSet` element."""
 
-    def __init__(self, fset, fcall=None, fapply=None):
+    def __init__(self, fset, fcall, vectorized=True):
         """Initialize a new instance.
 
         Parameters
         ----------
         fset : `FunctionSet`
             The set of functions this element lives in
-        fcall : `callable`, optional
+        fcall : `callable`
             The actual instruction for out-of-place evaluation.
             It must return an `FunctionSet.range` element or a
             `numpy.ndarray` of such (vectorized call).
-        fapply : `callable`, optional
-            The actual instruction for in-place evaluation.
-            Its first argument must be the
-            `FunctionSet.range` element
-            or the array of such (vectorization) to which the
-            result is written.
-
-        *At least one of the arguments fcall and fapply` must
-        be provided.*
+        vectorized : bool, optional
+            Whether the function supports vectorized evaluation
         """
         if not isinstance(fset, FunctionSet):
             raise TypeError('function set {!r} not a `FunctionSet` '
                             'instance.'.format(fset))
-
-        if fcall is None and fapply is None:
-            raise ValueError('call function and apply function cannot '
-                             'both be `None`.')
-
-        if fcall is not None and not callable(fcall):
-            raise TypeError('call function {!r} is not callable.'
-                            ''.format(fcall))
-
-        if fapply is not None and not callable(fapply):
-            raise TypeError('apply function {!r} is not callable.'
-                            ''.format(fapply))
+        if not callable(fcall):
+            raise TypeError('function {!r} is not callable.'.format(fcall))
 
         self._space = fset
-        self._call = fcall
-        self._apply = fapply
+        super().__init__(self._space.domain, self._space.range, linear=False)
+
+        # Determine which type of implementation fcall is
+        if isinstance(fcall, FunctionSetVector):
+            call_has_out, call_out_optional, _ = _dispatch_call_args(
+                bound_call=fcall._call)
+        elif isinstance(fcall, np.ufunc):
+            if fcall.nin != 1:
+                raise ValueError('cannot use `ufunc` with more than one '
+                                 'input parameter.')
+            call_has_out = call_out_optional = True
+        elif isfunction(fcall):
+            call_has_out, call_out_optional, _ = _dispatch_call_args(
+                unbound_call=fcall)
+        elif isclass(fcall):
```

Comment:
This could be implemented for `callable`, yes. However the dispatcher needs to be fenced by a `try...except` since not all `__call__` objects are inspectable (e.g. NumPy Ufuncs which are handled above and god knows which other objects).